recipes-app/iot2050-firmware-update/files/iot2050-firmware-update

```diff
@@ -168,14 +184,158 @@ class FirmwareUpdate(object):
+        base_addr_offset = cpuid_register_addr - base_addr
+        try:
+            f = os.open('/dev/mem', os.O_RDWR | os.O_SYNC)
+        except FileNotFoundError:
+            print("Open /dev/mem Failed")
+            sys.exit(1)
+
+        mem = mmap.mmap(f, mmap.PAGESIZE, mmap.MAP_SHARED, mmap.PROT_READ, offset=base_addr)
+        mem.seek(base_addr_offset)
+        data = []
+        data.append(struct.unpack('I', mem.read(4))[0])
+        device_id = hex(data[0])
+        current_cpu_id = int(device_id, 16) >> 11
+        if current_cpu_id not in self.board_cpu_ids:
+            print("Upgrade is not supported for the board w/ cpu id: {}".format(current_cpu_id))
+            sys.exit(1)
+
+        return self.board_cpu_ids[current_cpu_id]
+
+    @staticmethod
+    def _get_os_info() -> dict:
+        '''
+        Get the OS information by parsing the /etc/os-release
+
+        Returned is a dict that converted from /etc/os-release, for example:
+            NAME="debian"
+            VERSION_ID="3.1.1" 
+        
+        => 
+            {
+                "NAME": "debian"
+                "VERSION_ID": "3.1.1" 
+            }
+        '''
+        with open('/etc/os-release') as f:
+            return {l.split('=')[0]:l.strip().split('=')[-1].strip('"') for l in f.readlines()}
+
+class UpdateConfiguration(object):
+    def __init__(self, json_fileobj):
+        # Deserialize the json file to an object so that we can use dot operator to access the fields.
+        self._jsonobj = json.load(json_fileobj, object_hook=lambda d: Namespace(**d))
+
+    def _check_os(self, target_os, os_info) -> bool:
+        for tos in target_os:
+            if tos.key in os_info and os_info[tos.key].startswith(tos.version_prefix):
+                return True
+
+        return False
+
+    def firmware_filter(self, board_name: str, os_info: dict) -> list:
+        '''
+        Filter out the firmwares that could not be updated on the current board and current OS.
+
+        Return the list of the firmware names that could be updated.
+        '''
+        res = []
+        for firmware in self._jsonobj.firmware:
+            if firmware.target_board == board_name:
+                target_os = []
+                try:
+                    target_os = self._jsonobj.target_os
+                except AttributeError:
+                    pass
+
+                # local target_os configuration prior to global
+                try:
+                    target_os = firmware.target_os
+                except AttributeError:
+                    pass
+
+                # Only choose the firmware that have the target_os configuration identifying with current 
+                # OS info, or the firmware w/o target_os node which means it doesn't care the OS info.
+                if len(target_os) == 0 or self._check_os(target_os, os_info):
+                    res.append(firmware.name)
+
+        return res
+
+
+class UpdateControl(object):
```

Comment:
Would be nicer to make this a superclass of UpdateFirmware.
Maybe `FirmwareUpdate()` vs. `ManagedFirmwareUpdate(FimwareUpdate)`.

Reply:
Done, please review again.